src/processing/threshold.rs

```diff
@@ -0,0 +1,280 @@
+        let flat2 = flat.mapv(
+            |x| ((*x).to_f64().unwrap() * scale_factor).to_u8().unwrap()
+        );
+        let hist = flat2.histogram(grid);
+        // Straight out of wikipedia: 
+        let counts = hist.counts(); 
+        let total = counts.sum().to_f64().unwrap();
+        let counts = Array::from_iter(counts.iter());
+        // TODO: Could use the following for skimage-esque implementation 
+        // using CDFs 
+        // get cdf 
+        //let mut running_total = 0;
+        //let cdf = hist.counts().mapv(|x| {
+        //    running_total += x;
+        //    running_total as f32
+        //});
+        let mut sum_b = 0.0;
+        let mut weight_b = 0.0;
+        let mut maximum = 0.0;
+        let mut level = 0.0;
+        let mut sum_intensity = 0.0;
+        for (index, count) in counts.indexed_iter(){
+            sum_intensity += (index as f64) * (*count).to_f64().unwrap();
+        }
+        println!("Sum1: {:?}", sum_intensity);
+        for (index, count) in counts.indexed_iter(){
+            weight_b = weight_b + count.to_f64().unwrap();
+            sum_b = sum_b + (index as f64)* count.to_f64().unwrap();
+            let weight_f = total - weight_b;
+            if (weight_b > 0.0) && (weight_f > 0.0){
+                let mean_f = (sum_intensity - sum_b) / weight_f;
+                let val = weight_b * weight_f 
+			* ((sum_b / weight_b) - mean_f) 
+			* ((sum_b / weight_b) - mean_f);
+                if val > maximum{
+                    level = 1.0 + (index as f64);
+                    maximum = val;
+                }
+            }
+        }
+        threshold = level as f64 / scale_factor; 
+    }
+    Ok(threshold)
+}
+
+impl<T, C> ThresholdMeanExt<T> for Image<T, C>
+where
+    Image<T, C>: Clone,
+    T: Copy + Clone + Ord + Num + NumAssignOps + ToPrimitive + FromPrimitive + PixelBound,
+    C: ColourModel,
+{
+    type Output = Image<bool, C>;
+
+    fn threshold_mean(&self) -> Result<Self::Output, Error> {
+        let data = self.data.threshold_mean()?;
+        Ok(Self::Output {
+            data,
+            model: PhantomData,
+        })
+    }
+}
+
+impl<T> ThresholdMeanExt<T> for Array3<T>
+where
+    T: Copy + Clone + Ord + Num + NumAssignOps + ToPrimitive + FromPrimitive 
+{
+    type Output = Array3<bool>;
+
+    fn threshold_mean(&self) -> Result<Self::Output, Error> {
+        if self.shape()[2] > 1 {
+            Err(Error::ChannelDimensionMismatch)
+        } else {
+            let value = calculate_threshold_mean(&self)?;
+            let mask = apply_threshold(self, value);
+            Ok(mask)
+        }
+    }
+}
+
+fn calculate_threshold_mean<T>(array: &Array3<T>) -> Result<f64, Error>
+where
+    T: Copy + Clone + Num + NumAssignOps + ToPrimitive + FromPrimitive 
+{
+    Ok(array.sum().to_f64().unwrap() / array.len() as f64)
+}
+
+
+fn apply_threshold<T>(data: &Array3<T>, threshold: f64) -> Array3<bool>
```

Comment:
Why have threshold as `f64` and not type `T`?

Reply:
In general thresholding algorithms sometimes generate thresholds of a different type to the input (consider for example the mean threshold on `ints`, which at least from a theory perspective should be a float); if casting back to `T` is used, then we would need to decide on a strategy (`round` or `floor`?), which would likely depend also on whether the threshold value is used for `>` or `>=` comparison when being applied to generate the binary result. 

Am certainly open to adjusting this to `T` if we can decide on a consistent strategy that is also in-line with how this is done in other frameworks.

Reply:
That all makes sense, I'm happy with keeping it as `f64`